var/spack/repos/builtin/packages/py-python-lsp-server/package.py

```diff
@@ -0,0 +1,25 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyPythonLspServer(PythonPackage):
+    """A Python 3.7+ implementation of the Language Server Protocol."""
+
+    homepage = "https://github.com/python-lsp/python-lsp-server"
+    pypi = "python-lsp-server/python-lsp-server-1.6.0.tar.gz"
+
+    maintainers = ["alecbcs"]
+
+    version("1.6.0", sha256="d75cdff9027c4212e5b9e861e9a0219219c8e2c69508d9f24949951dabd0dc1b")
+
+    depends_on("python@3.7:", type=("build", "run"))
+    depends_on("py-setuptools@61.2.0:", type="build")
+    depends_on("py-jedi@0.17.2:0.19.0", type=("build", "run"))
+    depends_on("py-python-lsp-jsonrpc@1.0.0:", type=("build", "run"))
+    depends_on("py-pluggy@1.0.0:", type=("build", "run"))
+    depends_on("py-docstring-to-markdown", type=("build", "run"))
+    depends_on("py-ujson@3.0.0:", type=("build", "run"))
```

Comment:
Missing build/run dep on setuptools

Reply:
Not quite sure if I'm following. Do you mean that `py-setuptools` should be both a build and runtime dependency instead of just a build dep?

Reply:
Correct

Reply:
Gotcha! Should we change this in the default python template? I believe that it listed `py-setuptools` as a build only dependency and I defaulted to that when I uncommented the code and added the version constraint.

Reply:
Setuptools is rarely needed at run-time, but it is for this package.

Reply:
Ah okay interesting! For future reference how were you able to tell that setup tools was also a runtime dependency from looking at the package?

Reply:
There's a few ways you can be clued off to this. Things listed in `requires` under the `[build-system]` section of `pyproject.toml` or imported in `setup.py` are at a bare minimum a build dependency. If you also see them in `dependencies` under `[project]` in `pyproject.toml`, or in `install_requires` in `setup.py` or `setup.cfg`, then they are also needed at run-time. Oftentimes, developers will assume that setuptools is always installed and won't list it explicitly. If you recursively grep the source directory and see imports of `setuptools`, or more often `pkg_resources`, outside of `setup.py`, then it's also needed at run-time. Also, for older versions of Python, any usage of entrypoints also implies usage of setuptools at run-time. This is easier to see post-installation as there will be files in `prefix.bin` that didn't exist in the source dir but import setuptools/pkg_resources.

Reply:
Wow thanks for the detailed explanation! This is great! I hadn't thought of using grep to check for imports of setup tools throughout the project, I'll definitely have to try this out on future packages.

Reply:
Feel free to add this to the docs if you think it would help: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#setuptools